### ISSUE TITLE
Merging chunk maps into a single map via `Data.Typeable`

### ISSUE THREAD

**Issue by balacij** [JacquesCarette/Drasil#2873]
# Background
Presently, `ChunkDB`s are built as an enumeration of "chunk" maps;
https://github.com/JacquesCarette/Drasil/blob/10f9f9ed0e332c7a31875f6e035db96549c45945/code/drasil-database/lib/Database/Drasil/ChunkDB.hs#L165-L183

# Problem
i. We are unable to add (almost) arbitrary data into the ChunkDBs for various different systems.
ii. We manually track chunks in lists (by a single type, for later further processing) in `SystemInformation` when we already have them registered in the ChunkDB.
https://github.com/JacquesCarette/Drasil/blob/10f9f9ed0e332c7a31875f6e035db96549c45945/code/drasil-database/lib/Database/Drasil/SystemInformation.hs#L45-L53
iii. Adding arbitrary data to a ChunkDB is impossible without editing core `drasil-database` files, and difficult when editing core `drasil-database` files because it affects compilation of other examples.
iv. Working with similar maps requires redundant manual recreation (swhs and nopcm has a bit of this).
v. We have occurrences of conflicting UIDs because they are registered in different maps.

# Potential Solution : Merging Maps into a single one
I think we can solve these 5 or so problems by merging our maps into a single Map, and using more of the core `Data.Map` functionality.

We will need to hide type information to create a homogeneous Chunk type, and then merge our maps into a single Chunk Map.

The solution would, approximately, be as follows:
(1) Assume we have the following UID type, and typeclasses:
```haskell
type UID = String

-- | An interface to get the UID of a chunk.
class HasUID a where
    uid :: a -> UID

-- | Dump data to Strings for debugging-purposes.
class DrasilDumpable a where
    dump :: a -> String
```

(2) We build a `Chunk` type intended to carry arbitrary data in our final `ChunkDB`:
```haskell
import Data.Typeable -- this is an important design decision (as opposed to Type.Reflection) for simplicity
...
data Chunk where  -- Personally, I find GADTs to be more readable, but this can be done without them via ExistentialQuantification
    CHUNK :: (HasUID t, DrasilDumpable t, Typeable t) => t -> Chunk

instance HasUID         Chunk where uid  (CHUNK t) = uid t
instance DrasilDumpable Chunk where dump (CHUNK t) = dump t
...
```
Here, we build a commonality between all chunks registered in Drasil (e.g., they ALL must have a UID [a design decision], they all must be Typeable [Thanks to GHC 8.2+, this is the new 'normal'], and they all must be `DrasilDumpable` [not really required, but I think this would help in debugging -- this could also be something that we mass-dump into a single JSON file for our stable examples to show all raw knowledge we use to build up softifacts]). Additionally, we discard the real type information because the Typeable allows us to retrieve it later, and it will be helpful in creating a homogeneous map type.

(3) Thanks to capturing all of our chunks in a single Chunk type, we can construct a single homogeneous `ChunkDB` type:
```haskell
type ChunkDB = M.Map UID Chunk
```

(4) We build a few functions to help with registering chunks into the ChunkDB:
```haskell
-- Registering a list of like-typed chunks
chunksToChunkDB :: (HasUID t, Typeable t, DrasilDumpable t) => [t] -> ChunkDB
chunksToChunkDB = M.fromList . map (\x -> (uid x, CHUNK x))

-- Registering
registerChunk :: (HasUID t, Typeable t, DrasilDumpable t) => t -> ChunkDB -> ChunkDB
registerChunk c = M.insert (uid c) (CHUNK c)

-- Retrieval (I prefer using `Maybe` for prototyping instead of forced errors, but this would likely be a strict value or an error in practice)
retrieveChunk :: Typeable a => ChunkDB -> UID -> Maybe a
retrieveChunk cdb u = do
    (CHUNK r) <- M.lookup u cdb
    cast r
```

(5) When we want to pull arbitrary lists of chunks from the db by their `TypeRep`s, we can via casting:
```haskell
retrieveChunksByType :: (HasUID t, DrasilDumpable t, Typeable t) => ChunkDB -> TypeRep -> [t]
retrieveChunksByType cdb tr = relevantChunks
    where
        allChunks = M.toList cdb
        relevantChunks = mapMaybe (\(_, CHUNK c) -> if typeOf c == tr then cast c else Nothing) allChunks
```
NOTE: There is one problem here (which I go into small discussion into in my prototype), but it's not too important here. Essentially, the `if typeOf c == tr then cast c else Nothing` seems to be too strict of a requirement, and we might _want_ to lessen to just a `cast c` (in which case, we can get rid of the then redundant `TypeRep` parameter), but I'm not entirely certain yet (I will have to do more research on this).

Of course, whenever we want to do this, it will be rather inefficient, so we can create a cache:
(6) A Map of ChunkDBs by their TypeReps:
```haskell
type ChunksByTypeRep = M.Map TypeRep ChunkDB

dumpChunkDBToTypeRepMap :: ChunkDB -> ChunksByTypeRep
dumpChunkDBToTypeRepMap cdb = allRegistered
    where
        -- gather a list of registered types
        knownChunkTypes :: [TypeRep]
        knownChunkTypes = nub . map (typeOf . snd) $ M.toList cdb

        -- gather types with a list of all registered chunks that are of that type
        chunksWithTypes :: [(TypeRep, [Chunk])]
        chunksWithTypes = map (\x -> (,) x $ retrieveChunksByTypeInChunkBox cdb x) knownChunkTypes

        -- reconcile everything found, into a single ChunksByTypeRep map
        allRegistered :: ChunksByTypeRep
        allRegistered = M.fromList $ map (second chunksToChunkDB) chunksWithTypes
```

This still might not be the right solution for us because it constructs the `ChunksByTypeRepMap` after having fully constructed a single `ChunkDB`, but we can easily create an alternative solution which registers chunks in both maps when registering them.

```haskell
type ChunkDB' = (ChunkDB, ChunksByTypeRep)

-- There is one small problem here w.r.t uniqueness of UIDs and ensuring that the TypeRepMap is an accurate `view` of the `ChunkDB`
-- To fix it, we would just need to add an extra lookup to the initial chunkDB, removing it from the TypeRepMap manually before re-adding it to both maps (not a large problem, but I noticed it while typing up this ticket)
registerChunk' :: (HasUID t, Typeable t, DrasilDumpable t) => t -> ChunkDB' -> ChunkDB'
registerChunk' c (cdb, trcdb) = (cdb', trcdb')
    where
        u      = uid c
        chk    = CHUNK c
        cdb'   = M.insert u chk cdb
        trcdb' = M.alter
                    (maybe
                        (Just $ M.singleton u chk)
                        (Just . M.insert u chk))
                    (typeOf c)
                    trcdb

retrieveChunk' :: Typeable a => ChunkDB' -> UID -> Maybe a
retrieveChunk' (cdb, _) u = do
    (CHUNK r) <- M.lookup u cdb
    cast r

retrieveChunksByType' :: Typeable a => ChunkDB' -> TypeRep -> [a]
retrieveChunksByType' (_, trcdb) tr = maybe [] (mapMaybe ((\(CHUNK c) -> cast c) . snd) . M.toList) (M.lookup tr trcdb)
```

## Does it solve all noted problems?

i. We are unable to add (almost) arbitrary data into the ChunkDBs for various different systems.
    **Yes**, so as long all all relevant data is typeable, has a UID, and (optionally) can be dumped to a String (`DrasilDumpable`).

ii. We manually track chunks in lists (by a single type, for later further processing) in `SystemInformation` when we already have them in nicely placed maps.
    **Yes**, we would have additional functionality for grabbing all chunks in a chunkdb by a particular TypeRep. However, we lose out on the ability of polymorphic types with type constraint restrictions, but it seems that we stick to a small group of types anyway, so this ability may not be needed.

iii. Adding allowed datatypes to a ChunkDB is impossible without editing core `drasil-database` files, and difficult when editing core `drasil-database` files because it affects compilation of other examples.
    **Yes**, we will be solving both problems here because the data registerable is nearly arbitrary. Through this, `ChunkDB` should become _mostly_ stable and unchanging.

iv. Working with similar maps requires redundant manual recreation (swhs and nopcm has a bit of this).
    **Yes**, since `ChunkDB` is now a single type, we will be able to use `Data.Map`'s abilities to allow for easier merging of ChunkDBs and such.

v. We have occurrences of conflicting UIDs because they are registered in different maps.
    **Yes**. By merging in a single map, we ensure uniqueness is upheld in UIDs.

In general, I think this would decrease LOC while allowing us to generally get more out of our ChunkDBs.

## Notable issues
I can only think of 2 issues, but I'm unsure of how impactful they are.

1. We're using `Data.Typeable`. It appears that this library is largely unstable pre-GHC-8.2, and previously was designed differently. Additionally, since we are discarding type information and later re-gaining it, we will have _some_ amount of runtime impact, but I'm unsure of how much.
2. Retrieving chunks en masse by a single TypeRep is tricky, but I think we can sufficiently manage it. In (5), I briefly mentioned this. However, with a bit of hard work, we can mitigate the overhead -- we should be able to create functions similar to `makeLenses` which hides the difficulty associated with creating TypeReps and casting chunks/lists of chunks.

I'm posting this ticket earlier rather than later because I think it's a good proof of concept, but it still has problems. I don't think the problems are too 'breaking', but it would be best to discuss the base before I spend more time on this design.

# Prototype
I've made, and, lightly, tested this in my mini-development repo: https://github.com/balacij/ExprTests/blob/f7321d55611e7721a339188c79b66a3c6258ee8c/src/Lang/ChunkDB.hs

If you'd like to open up this code locally, I've currently setup `stack run` to run through a few small tests with this ChunkDB prototype.

**Comment by balacij** [comment]
I should also mention that the above code is just a prototype, I think we can refine it a bit more. In particular, I think we should be using `newtype`s for the ChunkDBs, and creating errors (either as an Either X String, or an actual plain `error`) where appropriate.

**Comment by JacquesCarette** [comment]
I think the technical ideas on the database side are sound. I'm going to want to work on the names of the functions provided by this, but that's not too hard. I am fine with the use of `TypeRep` and `Typeable`.

The thing that I am most curious about is how this will change (or enable change) in user code, as well as in `data-drasil`. My feeling is that if we adopt a monadic style (specifically: using a `State` monad with a `ChunkDB` state) for 'registering', the resulting code will likely be fine.  This 'registration' would be what replaces the explicit gathering into lists that we currently have. This is an important "second half" of doing this!

**Comment by balacij** [comment]
> I think the technical ideas on the database side are sound. I'm going to want to work on the names of the functions provided by this, but that's not too hard. I am fine with the use of TypeRep and Typeable.

This sounds great!

> The thing that I am most curious about is how this will change (or enable change) in user code, as well as in data-drasil. My feeling is that if we adopt a monadic style (specifically: using a State monad with a ChunkDB state) for 'registering', the resulting code will likely be fine. This 'registration' would be what replaces the explicit gathering into lists that we currently have. This is an important "second half" of doing this!

I like the idea of using `State` for 'registering'. I will need to spend a bit of time on this.

I also briefly considered how this would impact explicit need for gathering chunks into lists and then 'registering' them, but I was left a bit confused.

My first thought on how this "should be done" is: we wouldn't have any "collection" of chunks from within the modules they're defined. Instead, in the main executable program, we would perform some sort of reflection on known modules from the `lib` folders and packages, and gather all ones in their scopes. 

In Java, I would likely either use Java's Instrumentation API (Agents) to find the related modules at the "class loading time" (I'm not too sure what to call this; pre-runtime-loading?), or I would use Java's reflection lib (in the early stage of my programs runtime) to search for '@' Annotated definitions with a specific annotation class, and annotations that ensures the definition is kept for runtime usage & analysis (`@Retention(RetentionPolicy.RUNTIME)`, and `@Target(ElementType.TYPE)`).

I suppose the equivalent for the Instrumentation API could be a GHC plugin. Haskell has it's own reflection (and template haskell), which, _I think_ would have (at least) the same functionality as Java's.

Do you have any thoughts on how this second half should be done?

With the first method, we _might_ end up with less "boilerplate annotation" code placed in each "data" module.

**Comment by JacquesCarette** [comment]
We really don't want to do reflection in Haskell. A lot of this stuff in Java is really meta-linguistic (i.e. outside the actual language). I want to stay within Haskell as much as possible, and within well-understood Haskell too.

So we need to think about the process of registration. Doing inspection on an SRS to find used stuff is good. But we also need to "declare" which things from the standard databases (i.e. `data-drasil`) we're going to use. I don't want that part to be done via deep black magic.

**Comment by balacij** [comment]
>  We really don't want to do reflection in Haskell. A lot of this stuff in Java is really meta-linguistic (i.e. outside the actual language). I want to stay within Haskell as much as possible, and within well-understood Haskell too.
>  So we need to think about the process of registration. Doing inspection on an SRS to find used stuff is good. But we also need to "declare" which things from the standard databases (i.e. data-drasil) we're going to use. I don't want that part to be done via deep black magic.

Ah, ok. With your example, this makes a lot more sense to me now, thank you. I was thinking of a completely different direction!

>  Doing inspection on an SRS to find used stuff is good.

This sounds good to me.

>  But we also need to "declare" which things from the standard databases (i.e. data-drasil) we're going to use. I don't want that part to be done via deep black magic.

I wonder if we could be placing `ChunkDB`s inside `ChunkDB`s to indicate an imported "bundle of knowledge". Though, since UIDs for various chunks would be local to the `ChunkDB` they're placed in, we would need a new mechanism for UID generation/assignment. Manually written strings wouldn't quite cut it anymore because those strings would be local to the specific `ChunkDB`s they're intended to be placed in. I'm not quite sure how to resolve this yet (vaguely, it seems like we need to give `ChunkDB` context to chunk definitions when creating them). The sound of placing `ChunkDB`s within `ChunkDB`s sounds like a neat way of importing knowledge bundles however!

**Comment by balacij** [comment]
I think this ticket is also relevant to #2195, specifically in understanding the role of `SystemInformation`.

**Comment by JacquesCarette** [comment]
I don't necessarily want to put databases inside databases, but I think that creating a "current database for this example" from a collection of other databases, that's definitely a good idea.  Which is roughly what you arrived at in your EDIT.

As far as `UID`s are concerned, we can learn from URIs there, and use "qualified" names as a means to indicate provenance.

On foreground vs background: yes, but. What I mean is that I really want creators of an artifact to have to think about what they want in their artifact, then make an active choice. Now, when some sets of choice emerge as common, we can definitely create short-hands that automate it. [In many ways, a lot of doc-lang arose from such a process.]

And yes, that other ticket is relevant. The point is that we need to better understand the development process that we want for using Drasil. Because we've been so deep in prototype mode, we were mostly just doing things. Now is a good time to do more design. This is in part what the white paper is a prelude for.

The big question that underlies #2195 is just that: what's the end-to-end development process of a Drasil example? I mean, the process that we envision people **should** be using, when Drasil is able to support it.

**Comment by balacij** [comment]
> I don't necessarily want to put databases inside databases, but I think that creating a "current database for this example" from a collection of other databases, that's definitely a good idea. Which is roughly what you arrived at in your EDIT.
>  As far as UIDs are concerned, we can learn from URIs there, and use "qualified" names as a means to indicate provenance.

The main thing that I like about putting databases inside databases is that we get unique UIDs for cheap/free. However, it's quite complex as an embedded DSL, I haven't been able to come up with a design yet, and we'd likely need to be as careful as, if not more, general programming languages (the ones with support for modules & packages) in handling 'imports'. However, I might be thinking about this differently. Out of curiosity, do you think this is the main problem with databases in databases, or do you see a greater problem (or just other problems, I might just not be understanding how to design this at the moment)?

I think using qualified UIDs will mostly simplify the development. However, if we have conflicting UIDs for whatever reason, we might even see some Lens 'setter' usage to import specific modules with different UIDs to not cause breakage. Though, this might cause breakage in references from other imported chunks which depend on a re-UIDd chunk. I'm sure this will very rarely, if ever, occur, but, to mitigate it, it might be good for us to first build out the "concept domain" system you mentioned (to both analyze where breakage would occur, and fix it).

>  On foreground vs background: yes, but. What I mean is that I really want creators of an artifact to have to think about what they want in their artifact, then make an active choice. Now, when some sets of choice emerge as common, we can definitely create short-hands that automate it. [In many ways, a lot of doc-lang arose from such a process.]

Yes, I agree. I initially thought of coupling this kind of information with `CHUNK`s, but I quickly realized that it would be too specific to a specific definition of `background` and `foreground` (which might be different in each printer!). Thank you.

>  And yes, that other ticket is relevant. The point is that we need to better understand the development process that we want for using Drasil. Because we've been so deep in prototype mode, we were mostly just doing things. Now is a good time to do more design. This is in part what the white paper is a prelude for.
>  The big question that underlies #2195 is just that: what's the end-to-end development process of a Drasil example? I mean, the process that we envision people should be using, when Drasil is able to support it.

I see. I'll need to think about this for a while. Thank you :smile:

**Comment by balacij** [comment]
I've just now realized that I completely neglected to mention _why_ I looked into this. I've been looking into this because I need to be able to place arbitrarily parameterized types inside `ChunkDB`s and, potentially, `SystemInformation`s. I will need to keep prototyping to see the extents of what will be possible, and how.

**Comment by balacij** [comment]
> The thing that I am most curious about is how this will change (or enable change) in user code, as well as in data-drasil. My feeling is that if we adopt a monadic style (specifically: using a State monad with a ChunkDB state) for 'registering', the resulting code will likely be fine. This 'registration' would be what replaces the explicit gathering into lists that we currently have. This is an important "second half" of doing this!

I don't think I fully understand how/why we wouldn't need to gather into explicit lists anymore. Do you happen to have any examples of this?

**Comment by JacquesCarette** [comment]
```haskell
register :: (HasUID c, TypeRep c) => c -> ChunkDB -> ((), ChunkDB)
```
can be seen as
```haskell
do db <- get
     let db' = add c db
     put db'
```
aka
```haskell
modify (add c)
```
So we could gather pieces of `ChunkDB` instead. And wouldn't have to worry about types (i.e. homogeneous lists).

**Comment by balacij** [comment]
Ah, ok. Thank you.

Would we then have 1 large `[Chunk]` from which we use to insert all the chunks, or would we have 1 area of code that just has `modify`s called one after the other? Sorry, I'm still a bit confused as to how we can avoid the "explicit gathering".

**Comment by JacquesCarette** [comment]
There will be explicit gathering in `drasil-data`. There can be implicit gathering from SRS specifications.

Also, we don't want to call `modify` directly, we want to have our own functions (that may be thin wrappers).

**Comment by balacij** [comment]
Do you have a design in mind for the functions? If so, would it be easiest to do this in my first pass at this ticket?

As of right now, for a first pass at ticket, I'm assuming we're going to have a bunch of `Chunk` lists to replace the existing explicit lists.

Aside: I am working on this on the [`newChunkDB`](https://github.com/JacquesCarette/Drasil/tree/newChunkDB) branch.

**Comment by JacquesCarette** [comment]
No, I don't have such a design. I don't think it would be easiest to do it at the same time, as it's going to have serious repercussions for the examples.

**Comment by balacij** [comment]
`union` is a pleasant example of a `ChunkDB -> ChunkDB`. Import other ChunkDBs!

**Comment by balacij** [comment]
@JacquesCarette Do you know if there's any easy ways to debug <<loop>> errors?

**Comment by balacij** [comment]
Thank you. I think I will need to break this up a bit so that I can tackle the problems a bit more gradually. Something's going awry, and I'm struggling to find just where exactly it is.

Approximately, the next few steps are:
1. Replace all Chunk UID inheritances via "^. uid" and replace them with altered versions.
2. Understand what's going on with the high amount of `nw` usage near ChunkDB creation -- it seems like it should be done elsewhere, I'm not quite sure why we're pushing out things there.
3. Understand why we have manually set, hardcoded, and empty UIDs
4. Understand Trace vs Reference (vs Reference [e.g., Document Reference]), the comments are a bit difficult for me to read, but it seems like this system is built as a chunk where it's own UID is also used as the target object referenced UID. This would mean that if (1) goes well, this would still be a persistent problem. It would need to have a "target" UID stored as well.

**Comment by balacij** [comment]
The latest working version of this is posted on a [fresh repo](https://github.com/balacij/NewExprTests/blob/typeCheckedInDrasil/src/KnowledgeBase/ChunkDB.hs) I made specifically for testing out typing the expression language.

**Comment by balacij** [comment]
I'm not too sure there's much actionable for anyone on this ticket at the moment. The Haskell prototype worked well when I was using it. This ticket is currently blocked by, in some capacity, resolving #2911.

UID collisions would occur if we were to merge the maps _today_. I started changing some UIDs (too early!) in #2904, but we then moved it back into discussion in #2911, which is where we're currently blocked. So, I think #2911 is the one we need to give our attention to the most.

**Comment by balacij** [comment]
Before code gets lost, I posted the latest copy of the "merged ChunkDBs" to an external repository: https://github.com/balacij/ProtoChunkDB

There is a branch on here (Drasil repo) where I started the conversion of our ChunkDBs into the new ChunkDBs, but the progress is largely halted due to issues with UID "conflicts" (continued in #2911).

**Comment by balacij** [comment]
One future work idea: "chunk holes." Think about "type-hole programming" but with the data entered into chunks. These are chunks that we might always prefer to build "later on" in the development of a case study (e.g., I think we did this with ODEInfo for ODEs before @cd155 came along?) or for creating placeholders during the development of things, without creating temporary/mock data.

**Comment by balacij** [comment]
I _believe_ that with the existing PRs enabling `StrictData` (mentioned in https://github.com/JacquesCarette/Drasil/issues/3710#issuecomment-2688582864) up to and including `drasil-database`, the "only issue" (which is nowhere nearly as simple as the `StrictData` issue is) blocking this PR is now the UID conflicts. The only package that we don't `StrictData` working for yet is `drasil-gool`, but that's an active issue (https://github.com/JacquesCarette/Drasil/issues/3710#issuecomment-2689246630, and definitely not one that I think is as high priority as this issue is).

**Comment by balacij** [comment]
Returning to this looking for the next actionable step for implementing the prototype 😓

<details>

<summary>Example of the Chunk conflicts from `glassbr`.</summary>

Format:
```json
{ ...,
  "ChunkUID: [List of chunk types instantiated with the same ChunkUID],
  ...
}
```


```json
{
   ...
    "doc:TraceGraphAvsAll": [
        "references",
        "labelledContent"
    ],
    "minThick": [
        "terms",
        "symbols",
        "references",
        "labelledContent",
        "dataDefinitions"
    ],
    "pascal": [
        "units",
        "terms"
    ],
    ...
    "stdOffDist": [
        "terms",
        "symbols",
        "references",
        "labelledContent",
        "dataDefinitions"
    ],
    "sysSetValsFollowingAssumps": [
        "references",
        "concepts",
        "conceptInstances"
    ],
   ...
    "theory:isSafeLoad": [
        "theoryModels",
        "references",
        "labelledContent"
    ],
   ...
    "theory:riskFun": [
        "references",
        "labelledContent",
        "instanceModels"
    ],
    ...
    "varValsOfmkE": [
        "references",
        "concepts",
        "conceptInstances"
    ],
   ...
    "x": [
        "terms",
        "symbols"
    ],
   ...
}

```

</details>


There are some common conflicts:
1. _`references` with almost every other data type._
2. _`concepts`_ with _`conceptInstances`._
3. _`terms`_ & _`symbols`._
4. _(data definitions, instance models, theory models, general definitions) & (terms, symbols)._
5. _`units` & `terms`._

The references look like the most obvious first target. Everything else will require a bit more thinking about the chunk graph. Discussion of references to follow.

### References

<details>

<summary>Snippet from the chunk dump tables:</summary>


```
Reference:
----------------------------------------------------------------------------------------------------
UID                           Reference Address             Short Name
----------------------------------------------------------------------------------------------------
AppendixLabel                 Sec:Appendix                  Appendix
....
TraceGraphAllvsAllLink        ../../../../traceygraphs/glassbr/allvsall.svg
                                                            TraceGraphAllvsAll
...
aR                            DD:aspectRatio                aspectRatio
accAlteredGlass               accAlteredGlass               Accommodate-Altered-Glass
...
astm2016                      astm2016                      astm2016
beasonEtAl1998                beasonEtAl1998                beasonEtAl1998
checkGlassSafety              checkGlassSafety              Check-Glass-Safety
checkInputWithDataCons        checkInputWithDataCons        Check-Input-with-Data_Constraints
glassBRSRSLink                https://github.com/smiths/caseStudies/tree/master/CaseStudies/glass
                                                            glassBRSRSLink
glassTypeFac                  DD:gTF                        gTF
...
maintainable                  maintainable                  Maintainability
...
doc:ToU                       Table:ToU                     ToU
doc:TraceGraphAllvsAll        Figure:TraceGraphAllvsAll     TraceGraphAllvsAll
...
theory:dimlessLoad            IM:dimlessLoad                dimlessLoad
...
```

</details>

We have a real mixed bag of items, but it looks like two major categories of "references":
1. Document-internal references.
2. References to external data.

The documentation of references claims a bit different:

```haskell
-- | Applying different pieces of information for a reference.
-- An RP is a decorated internal reference.
-- Citation is a citation.
-- URI is for URLs and other external links.
data LblType =
    RP IRefProg String
  | Citation String
  | URI String
```

#### Discussion

Some aside commentary:
1. While all of these things make sense, I think I prefer to think of things as "Drasil-generated and -understood things" and "non-Drasil-generated and -unknown things". A citation-type reference is still a document-internal reference. Looking at what data is contained in a Reference sort of evidences my point. While "URI" is strictly correct, I think it's better to think of this as "non-Drasil-generated things", for which we happen to use URIs to reference.
2. Across document formats, data that encodes references to purely external data (e.g., files, websites, sockets, etc.) should always be the same. The internal document references are different across file types. For example, an internal document reference in Java is probably an absolute reference containing all of: package name + class name + member name + type signature. An internal document reference in LaTeX and HTML are IDs, but the ID conventions are quite different. Currently, our IDs are shared in LaTeX and HTML are the same.

So, where do all of these references come from? We have two kinds of references at the moment:
1. Generated ones, for almost all of our chunks.
2. Manually created ones, for our external data.

The generated ones are created through:

https://github.com/JacquesCarette/Drasil/blob/e1790c75d6ab323a0afcab9bcd0cc9d04bb6bfed/code/drasil-docLang/lib/Drasil/DocumentLanguage.hs#L151-L183

Now, I like this! But I don't think they should always be created for all of our chunks. They should only be created if we render them in the documents (which, yes, happens most of the time).

The manually created ones are created often as follows:

https://github.com/JacquesCarette/Drasil/blob/e1790c75d6ab323a0afcab9bcd0cc9d04bb6bfed/code/drasil-example/dblpend/lib/Drasil/DblPend/Body.hs#L185-L190

The last 4 lines create the external-reference chunk, and the first line uses it. I do think that this external-reference chunk should be a bit more insightful, carrying information about things like "alternative text" (e.g., answering "what is it?"), checksums (e.g., "what should it be?"), etc. So, I kind of think that an external-reference chunk is a bit more special than an internal one. It should probably be split off. I'm not saying that an internal reference can't be as rich as an external one, but they're rich in different ways.

#### Thoughts

Finally, some more actionable thoughts:
1. To avoid UID collisions, I think we need to do two things:
   1. Demote internal-kind `Reference` to something other than a chunk. It's display knowledge. I think it should be demoted to intermediate data that the writer would create on-the-fly.
   2. Promote external-kind `Reference` to a full chunk.
2. We've had some related discussions to both of these points previously as well. I'll need to find the links later.
3. Short-term, to avoid UID collisions, this can be moved out of `ChunkDB`.
4. Long-term, I think we should consider rebuilding them.

**Comment by balacij** [comment]
Thank you. There's a lot more to discuss regarding references.

Some brief notes on what I think should happen next too:
* I think `terms` and $X$ are the next target. Terms are either extended with $X$ (i.e., with richer information, such as symbol, unit, theory, etc.) or floating terms with only simple information (i.e., no $X$, only name, short name, and description).
* `Concept` and `ConceptInstances` next or at the same time too. I haven't looked into this one enough yet.
* On `LabelledContent`, I kind of want to say that this is just an external reference too. It's just that it's one where we want to try to inline some of the data from the external reference. Whether it be a figure, a table, etc. However, we currently have some `LabelledContent` like the assumptions lists (e.g., our recent PR #4012) that need to be removed altogether in favour of richer chunks that are rendered 'correctly', with a corresponding table, etc.

**Comment by balacij** [comment]
https://github.com/JacquesCarette/Drasil/blob/f8e88e58c964af867c62f70c445e9ec3e5de5b11/code/drasil-theory/lib/Data/Drasil/TheoryConcepts.hs#L7-L17

For each of our 'model kinds' (for lack of better words), we create a `CI` that discusses the word itself (which we use when referencing $x$Models abstractly), and then we have our actual Haskell-level types that give them structure we can work with operationally.

Two things:
1. With the eventual upgraded `ChunkDB`, I think we can generate these `CI`s if they aren't already created on the first time a chunk type is registered. Perhaps excluding the acronyms, which I don't think belong in a `CI` anyways (I think we discussed this in the past?).
2. The `CI`s are "meta-knowledge" (which probably belong in `drasil-metadata`). 

I'm having a hard time taking my eyes off (2) for some reason. What's the deeper meaning here?

Considering "Drasil in Drasil", are we seeing the English description of the hypothetical chunk that captures "$X$Model" rendered as a chunk and the computational knowledge/axiomatic knowledge rendered as Haskell code?

Maybe I'm just tired and I'm seeing things that aren't there? Not sure, but I want to note this down and stop wasting time.

**Comment by balacij** [comment]
So that I don't forget what I'm currently working on: As in #4022, we're "cheating" with our chunk gathering of `UnitDefn`s by gathering (in fact, we're not just gathering them, but we're generating duplicate copies seemingly mismatching information [??? needs further investigation]) them retroactively from chunks that are already in the `ChunkDB`. We need to stop that.

**Comment by JacquesCarette** [comment]
There is a lot of actionable information in this issue. I think going through it all could form the basis of another design meeting, in particular all the [chunk conflicts](https://github.com/JacquesCarette/Drasil/issues/2873#issuecomment-2693129320). Resolving these requires real design + understanding, but the implementation might make good material for people who are new to Drasil.

**Comment by balacij** [comment]
Does `ChunkDB` want to be a (flattened) theory?

I can vaguely see how this makes sense:
1. An `IdeaDict` approximately translates to a "constant symbol." We previously differentiated between "vocabulary," "function," and "relation" symbols, but I think an `IdeaDict` would translate to the union of all 3. Is that an issue?
2. A `ConceptChunk` approximately translates to "(vocabulary|function|relation) constants definition in the ontological dimension." In other words, defining English descriptions of the (vocabulary|function|relation) constants.

(I'll explain rationale in a follow-up comment.)

What could this say about the design of `ChunkDB`, `IdeaDict`, `ConceptChunk`, the closely related `ConceptInstance`, and all other chunks? Thoughts:
1. `IdeaDict` would be erased. A simple cache of known `NP`s would instead be carried in the `ChunkDB`.
2. `ConceptChunk` would also be erased. As a reminder, the [`ConceptChunk`](https://github.com/balacij/DrasilInvestigation/blob/main/Designing%20Drasil's%20Chunks%202025/Analysis/Chunk%20Types/ConceptChunk.md) only extends `IdeaDict`s with English descriptions (`Sentence`s) and ["concept domain"](https://github.com/balacij/DrasilInvestigation/blob/main/Designing%20Drasil's%20Chunks%202025/Analysis/Typeclasses/ConceptDomain.md) information.
    1. The English descriptions would be replaced with two things. The first is a requirement that all chunks inserted into the `ChunkDB` must be of a type that satisfies a hypothetical typeclass with at least the following field: `toHumanReadableDescription :: a -> Sentence` (Yes, the name needs work.). Then, on insertion of the chunk into the `ChunkDB`, we would automatically extract the corresponding human-readable description of the structured knowledge we're encoding with Drasil. (Meta: I think this means that chunk instances are instances of axiomatic knowledge?)
    2. For the few things for which we don't have a real, structured chunk for, but which we give an English definition to, we need to create an `OpaqueDefinition` chunk type that contains only a `Sentence` (with a single type parameter that indicates the chunk type the definition should have actually taken on if we had an actual definition). (Meta: This is where things start to break down -- What would those types be? Would those types be instantiable? For example, "software engineering" is a suitable candidate for an `OpaqueDefinition Domain` type signature. Giving a complete definition of any domain would be impossible, so maybe `Domain` would be uninstantiable [i.e., no constructors]. The point of this would be that we know that something is a term, that it has a human-readable [intensional] definition, but that the mathematical [extensional] definition would be impossible [or just too difficult] to write. I'll assume this is an okay idea for now, but I'm hesitant. This needs to be fleshed out further.)
3. Our `ConceptInstance`s would all be replaced with either `OpaqueDefinition`s or our 'real' chunks. Recalling Projectile's [`ConceptInstance`s](https://github.com/balacij/DrasilInvestigation/tree/main/Designing%20Drasil's%20Chunks%202025/Analysis/Chunks/ConceptInstances), all of them are assumptions. Actually, we also have (goals and [non-]functional requirements)[https://github.com/JacquesCarette/Drasil/blob/5705da6c270bd1839801037c71a5e8466072e655/code/drasil-example/projectile/lib/Drasil/Projectile/Body.hs#L190-L191]. Note: I didn't add those to the `DrasilInvestigation` Obsidian repo because they were only inserted into the `ChunkDB` -- they were not directly referenced by the SI (!!!!!). I've already stated this somewhere, and I think we've already discussed this, but all of our [assumptions carry some sort of axiomatic knowledge](https://jacquescarette.github.io/Drasil/examples/projectile/SRS/HTML/Projectile_SRS.html#Sec:Assumps) in them that just hasn't been structured yet (e.g., the definitions of constants [$g=9.8$] and the definitions of symbols [$v_i=0$, $p_i=(0,0)$]). Note that some are difficult to describe axiomatically (e.g., `freeFlight: The flight is free; there are no collisions during the trajectory of the projectile.` and `neglectCurv: The distance is small enough that the curvature of the celestial body can be neglected.`) however. (Meta: For the most part, this is okay, but the last difficult with transcribing some assumptions could be an issue. Then again, they could just be `OpaqueDefinition`s. Though, what would the type parameter be filled in with? Is this where Haskell is an issue or where our understanding of how to use Haskell and/or what we're building is breaking down?)
4. See (2.1). Not much else that I can think of right now 😄.

So, this would bring about some good changes to `ChunkDB` and co. that I already wanted really (so maybe I'm biased, I'll have to argue rationale in the follow-up further). Specifically, I'm talking about:
1. The hypothetical typeclass containing `toHumanReadableDescription`.
2. The automatic gathering of terms.
3. The automatic gathering of human-readable definitions of terms.
4. The added requirement that all inserted chunks need to have human-readable descriptions.

(4) and (4) here are strongly related. This is further related to our discussion on Monday (balacij/Doctorate#83) where we agreed that there should be a duplicate definitions for terms for different "dimensions" (ontological vs axiomatic). It would mean that at the "theoretic" level, chunks correspond to "(function|relation) symbol definitions" ("chunk ≈ a line item in a theory"). Does this provide insight into how we should be using Haskell to build Drasil? Probably, but I'm not sure yet. It's still muddy. I'm still fairly convinced that we're dealing with a lot of accidental complexities working with Haskell instead of our idealized Drasil language that blurs vision.

Now, three questions:
1. Where do `UID`s fit? They're a key component of `ChunkDB` after all!
2. Where did `ConceptDomain` go? I forgot to discuss it!
3. Do we want to permit type parameters for chunks insertable into the `ChunkDB`? What do they really mean at the ontology level? Subtype? For `ModelKinds`, they indicate "usable in code generation" depending on `Expr`/`ModelExpr`.

Thoughts on said questions (both of which, please preface with "I'm not sure! But maybe..."):
1. They stay, but we gained some insight into them. Assuming "chunk ≈ a line item in a theory," they could really refer to the human-readable names we give items in a theory (I'm thinking of presentations from Dr Farmer's book), or they could also give insight into `DSentence`/how that chunk came to be. They could also just be an implementation wart, in which case we could just replace them with a hash code of the hypothetical `DSentence`.
2. They get deleted. `ConceptDomain`s are only functionally used for sorting [non-]functional requirements, goals, and assumptions for use in the SRS renderer. However, as I mentioned earlier, these should all be replaced with either "real chunks" or `OpaqueDefinition`s. All other times we fill in `ConceptDomain` lists, we will it with "domains that this chunk instance belongs to." For example, physics, software engineering, etc. None of this is used yet, but it is very important knowledge. I'm not sure how this information could/should be used, but it's information that should exist somehow. To be frank, I'm not sure how this information is recreated. My initial thinking is that they all translate to references to `OpaqueDefinition`s for now while we try to figure that out. For some things, it _might_ be something that we can trace/calculate using the theory lattice we use to build the initial theory that a Drasil-based software project begins with (Note: we discussed this in passing, but the "Characteristics of the intended reader" section gives the boundary information of the initial theory [what it should contain/shouldn't], so it might be something that we can trace to the theory lattice -- further assuming we have a theory lattice that builds our fundamental terms and such used in an SRS/initial theory, which I don't think is all that crazy of an assumption). For what "physics" and other domains would have a type of, it would have to be for a data type (e.g., `Domain`) with no constructors.
3. I tend towards 'yes' but only for operational purposes. It gives us constraints that GHC can statically assert for us. Building Drasil in Drasil, they would pose an issue, however, because we use them to say different relations to other types.

Finally, (2) brings about an issue: we need to encode "domain" as an `OpaqueDefinition`. How would we do that? `OpaqueDefinition ()`? Well, that _works_ but it's cheating because we don't have a conceptual understanding of how/why this works and we don't want to hack things into Drasil. Well, this also brings about a greater issue: we don't enforce that Drasil contains (the equivalent of) `OpaqueDefinition`s for all of our chunk types. We do manually create `ConceptChunk`s for `TheoryModel`, `InstanceModel`, etc., but we definitely don't create it for all of our chunk types. This should probably be a requirement for all of our chunk types (a bit of `Proxy`-magic with typeclasses and I think we can create a static typeclass-based constraint).

**Comment by balacij** [comment]
Strictly speaking, the original issue is resolved. I don't see any notes here that are worth keeping (or at least would be difficult to find should we need to). We now have some resultant work:

* https://github.com/JacquesCarette/Drasil/issues/5471
* https://github.com/JacquesCarette/Drasil/issues/5474
* https://github.com/JacquesCarette/Drasil/issues/5450
* https://github.com/JacquesCarette/Drasil/issues/5449
* https://github.com/JacquesCarette/Drasil/issues/5168
* https://github.com/JacquesCarette/Drasil/issues/5054
* https://github.com/JacquesCarette/Drasil/issues/4541
* https://github.com/JacquesCarette/Drasil/issues/4150
* https://github.com/JacquesCarette/Drasil/issues/2908
* https://github.com/JacquesCarette/Drasil/issues/4300
* https://github.com/JacquesCarette/Drasil/issues/4671
